crates/numcodecs-python/src/schema.rs

```diff
@@ -0,0 +1,486 @@
+use std::{
+    borrow::Cow,
+    collections::{hash_map::Entry, HashMap},
+};
+
+use pyo3::{intern, prelude::*, sync::GILOnceCell};
+use pythonize::{depythonize_bound, PythonizeError};
+use schemars::Schema;
+use serde_json::{Map, Value};
+use thiserror::Error;
+
+use crate::PyCodecClass;
+
+macro_rules! once {
+    ($py:ident, $module:literal $(, $path:literal)*) => {{
+        fn once(py: Python) -> Result<&Bound<PyAny>, PyErr> {
+            static ONCE: GILOnceCell<Py<PyAny>> = GILOnceCell::new();
+            Ok(ONCE.get_or_try_init(py, || -> Result<Py<PyAny>, PyErr> {
+                Ok(py
+                    .import_bound(intern!(py, $module))?
+                    $(.getattr(intern!(py, $path))?)*
+                    .unbind())
+            })?.bind(py))
+        }
+
+        once($py)
+    }};
+}
+
+pub fn schema_from_codec_class(
+    py: Python,
+    class: &Bound<PyCodecClass>,
+) -> Result<Schema, SchemaError> {
+    if let Ok(schema) = class.getattr(intern!(py, "__schema__")) {
```

Comment:
This magic constant and others (`_ty`, check for all interns) should be internal consts to ease refactoring